source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -147,23 +226,124 @@ void CacheFilter::onBody(Buffer::InstancePtr&& body) {
+}
+
+void CacheFilter::injectValidationHeaders() {
+  ASSERT(lookup_result_, "injectValidationHeaders precondition unsatisfied: lookup_result_ "
+                         "does not point to a cache lookup result");
+  ASSERT(validating_cache_entry_, "injectValidationHeaders precondition unsatisfied: the "
+                                  "CacheFilter is not validating a cache lookup result");
+  ASSERT(request_headers_, "injectValidationHeaders precondition unsatisfied: request_headers_ "
+                           "does not point to the RequestHeadersMap of the current request");
+  const Http::HeaderEntry* etag_header =
+      lookup_result_->headers_->get(Http::CustomHeaders::get().Etag);
+  const Http::HeaderEntry* last_modified_header =
+      lookup_result_->headers_->get(Http::CustomHeaders::get().LastModified);
+
+  if (etag_header) {
+    absl::string_view etag = etag_header->value().getStringView();
+    request_headers_->setReferenceKey(Http::CustomHeaders::get().IfNonMatch, etag);
+  }
+  if (last_modified_header) {
+    absl::string_view last_modified = last_modified_header->value().getStringView();
+    request_headers_->setCopy(Http::CustomHeaders::get().IfUnmodifiedSince, last_modified);
+  }
 }
+
+void CacheFilter::encodeCachedResponse() {
+  ASSERT(lookup_result_, "encodeCachedResponse precondition unsatisfied: lookup_result_ "
+                         "does not point to a cache lookup result");
+
+  response_has_trailers_ = lookup_result_->has_trailers_;
+  const bool end_stream = (lookup_result_->content_length_ == 0 && !response_has_trailers_);
+  // TODO(toddmgreer): Calculate age per https://httpwg.org/specs/rfc7234.html#age.calculations
+  lookup_result_->headers_->addReferenceKey(Http::Headers::get().Age, 0);
+
+  // Set appropriate response flags and codes
+  Http::StreamFilterCallbacks* callbacks_ =
+      filter_state_ == FilterState::Encoding
+          ? static_cast<Http::StreamFilterCallbacks*>(encoder_callbacks_)
+          : static_cast<Http::StreamFilterCallbacks*>(decoder_callbacks_);
+
+  callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::ResponseFromCacheFilter);
+  callbacks_->streamInfo().setResponseCodeDetails(
+      CacheResponseCodeDetails::get().ResponseFromCacheFilter);
+
+  // If the filter is encoding, 304 response headers and cached headers are merged in encodeHeaders
+  // If the filter is decoding, we need to serve response headers from cache directly
+  if (filter_state_ == FilterState::Decoding) {
+    // decoder_callbacks_->encodeHeaders() will invoke CacheFilter::encodeHeaders, skip it
+    skip_encoding_ = true;
+    decoder_callbacks_->encodeHeaders(std::move(lookup_result_->headers_), end_stream);
+  }
+
+  if (lookup_result_->content_length_ > 0) {
+    remaining_body_.emplace_back(0, lookup_result_->content_length_);
+    getBody();
+  } else if (response_has_trailers_) {
```

Comment:
Double-checking, may well be correct as-is: if lookup_result_ has content_length>0, you never want to check for trailers at the same time?

Reply:
When `getBody` is called and the body is fetched from cache, the callback `CacheFilter::onBody` is invoked. At the end of this, a check is performed for more body data / trailers.